api/python/t4/data_transfer.py

```diff
@@ -607,3 +610,120 @@ def _process_url(src):
             results = executor.map(_process_url, src_list)
 
     return results
+
+
+def select(url, query, query_type="SQL", meta=None, s3client=None, raw=False, **kwargs):
+    accepted_compression = {
+        '.bz2': 'BZIP2',
+        '.gz': 'GZIP'
+        }
+    format_compression = {
+        'Parquet': ['NONE'],  # even if column-level compression has been used.
+        'JSON': ['NONE', 'BZIP2', 'GZIP'],
+        'CSV': ['NONE', 'BZIP2', 'GZIP'],
+        }
+    # largely 1:1, but we still need to check that they're accepted formats.
+    ext_formats = {
+        '.parquet': 'parquet',
+        '.json': 'json',
+        '.jsonl': 'jsonl',
+        '.csv': 'csv',
+        '.tsv': 'tsv',
+        }
+    valid_s3_select_formats = {
+        'parquet': 'Parquet',
+        'json': 'JSON',
+        'jsonl': 'JSON',
+        'csv': 'CSV',
+        'tsv': 'CSV',
+        }
+
+    parsed_url = urlparse(url)
+
+    bucket, path, version_id = parse_s3_url(parsed_url)
+
+    # TODO: Use formats lib for this stuff
+    # use metadata to get format and compression
+    compression = None
+    format = meta.get('target')
+    if format is None:
+        format = meta.get('format', {}).get('name')
+        if format in ('bzip2', 'gzip'):
+            compression = format.upper()
+            format = meta.get('format', {}).get('contained_format', {}).get('name')
+
+    # use file extensions to get compression info, if none is present
+    exts = pathlib.Path(path).suffixes
+    if exts and not compression:
+        if exts[-1].lower() in accepted_compression:
+            compression = accepted_compression[exts.pop(-1)]
+    compression = compression if compression else 'NONE'
+
+    # use remaining file extensions to get format info, if none is present
+    if exts and not format:
+        ext = exts[-1].lower()
```

Comment:
How about making ext not include '.' and removing it from the values in the list, since you always strip it anyway.

Reply:
I did that because I don't strip it from any of the extensions, and it's more visually clear that it's an extension name (and not a format name) in the dict -- also, it skips the strip('.') calls.

...

Oh, I think I see.  That `[-1]` isn't stripping the '.' from the extension, it's selecting the last extension in the list of extensions (as in the case of `'.csv.gz'` or `.json.bz2`.  Compression extension (if any) has already been removed, and that negative reference handles cases where someone names their file `"blah foo 2009.10.19.json.gz"` or something similarly goofy -- then 'path.suffixes' would at this point be `['.10', '.19', '.tar']`.

Added comments for clarity, let me know if that's sufficient.